cuqi/mcmc/_sampler.py

```diff
@@ -0,0 +1,324 @@
+from abc import ABC, abstractmethod
+import os
+import sys
+import numpy as np
+import pickle as pkl
+import cuqi
+from cuqi.samples import Samples
+from progressbar import progressbar
+
+class SamplerNew(ABC):
+    """ Abstract base class for all samplers.
+    
+    Provides a common interface for all samplers. The interface includes methods for sampling, warmup and getting the samples in an object oriented way.
+
+    Samples are stored in a list to allow for dynamic growth of the sample set. Returning samples is done by creating a new Samples object from the list of samples.
+
+    """
+
+    def __init__(self, target: cuqi.density.Density, initial_point=None, callback=None):
+        """ Initializer for abstract base class for all samplers.
+        
+        Parameters
+        ----------
+        target : cuqi.density.Density
+            The target density.
+
+        initial_point : array-like, optional
+            The initial point for the sampler. If not given, the sampler will choose an initial point.
+
+        callback : callable, optional
+            A function that will be called after each sample is drawn. The function should take two arguments: the sample and the index of the sample.
+            The sample is a 1D numpy array and the index is an integer. The callback function is useful for monitoring the sampler during sampling.
+
+        """
+
+        self.target = target
+        self.callback = callback
+
+        # Choose initial point if not given
+        if initial_point is None:
+            initial_point = np.ones(self.dim)
```

Comment:
Just a tiny reminder/discussion here for future work with LinearRTO: since LinearRTO is not a Markov method, (maybe) it doesn't need an `initial_point`? and should `np.ones(self.dim)` be considered as an independent sample for LinearRTO?

Reply:
#378